route_guide/route_guide_server.cc

```diff
@@ -55,166 +66,259 @@ float GetDistance(const Point& start, const Point& end) {
+               Point, // Request from client.
+               Feature>( // Response from the server.
+               "GetFeature")
+        | Map(Then([this](auto&& context) {
+             // The Server::Handler() can be composed with other tasks using
+             // the "|" operator.
+             // Its .body() callback is invoked every time the client sends a
+             // request to the server.
+             return Server::Handler(std::move(context))
+                 .body([this](auto& call, auto&& point) {
+                   // Check if the request is valid.
+                   if (!point) {
+                     return;
+                   }
+                   Feature feature;
+                   // Populate the response message with data.
+                   feature.set_name(
+                       GetFeatureName(*point.get(), feature_list_));
+                   feature.mutable_location()->CopyFrom(*point.get());
+                   // Send the response to the client and finish the call.
+                   call.WriteAndFinish(feature, Status::OK);
+                 });
+           }))
+        | Loop();
+  }
 
-    status = server->Serve<RouteGuide, Rectangle, Stream<Feature>>(
-        "ListFeatures",
-        [this](auto* call, auto&& rectangle) {
-          auto lo = rectangle->lo();
-          auto hi = rectangle->hi();
-          long left = (std::min)(lo.longitude(), hi.longitude());
-          long right = (std::max)(lo.longitude(), hi.longitude());
-          long top = (std::max)(lo.latitude(), hi.latitude());
-          long bottom = (std::min)(lo.latitude(), hi.latitude());
-          for (const Feature& f : feature_list_) {
-            if (f.location().longitude() >= left &&
-                f.location().longitude() <= right &&
-                f.location().latitude() >= bottom &&
-                f.location().latitude() <= top) {
-              call->Write(f);
-            }
-          }
-          call->Finish(Status::OK);
-        },
-        [](auto*, bool) {});
-
-    if (!status.ok()) {
-      return status;
-    }
+  // ServeListFeatures() function demonstrates a server streaming RPC. The
+  // client sends Rectangle messages, the server responds with a stream
+  // of addresses that fall within the Rectangle's area.
+  auto ServeListFeatures(Server* server) {
+    return server->Accept<
+               RouteGuide,
+               Rectangle,
+               Stream<Feature>>("ListFeatures")
+        | Map(Then([this](auto&& context) {
+             // The Server::Handler() can be composed with other tasks using
+             // the "|" operator.
+             // Its .body() callback is invoked every time the client sends a
+             // request to the server.
+             return Server::Handler(std::move(context))
+                 .body([this](auto& call, auto&& rectangle) {
+                   // Check if the request is valid.
+                   if (!rectangle) {
```

Comment:
As with the comment below about type-safe checks:

```suggestion
                   if (rectangle == nullptr) {
```